### PR TITLE
Cgroup monitoring and alerting

### LINKED ISSUE #16392: Cgroup monitoring and alerting
### Is your feature request related to a problem? Please describe:

There isn't yet a mechanism to monitor the cgroup cpu/memory limit and do corresponding alerting.

### Describe the feature you'd like:

Add a monitor to watch cgroup cpu/memory limit setting changes and update corresponding metrics. Display the limits on related Grafana graphs. Trigger alerting when cpu/memory usage exceeds a threshold relative to the limit.

### Describe alternatives you've considered:

None.

### Teachability, Documentation, Adoption, Migration Strategy:

In real scenarios, users like to change cpu/memory limits online w/o restarting the service, show the current limit somewhere and receive alerting when the usage is approaching the limit.

### PR DESCRIPTION
**Pull request by pingandb**
### What is changed and how it works?

Issue Number: Close #16392

What's Changed:
A cgroup monitor background worker task was added to periodically watch the cgroup cpu/memory limit setting changes. Metrics will be set accordingly and displayed on Grafana. New alerting rules were also added for the cpu/memory usage ratio relative to the limits.

```
A cgroup monitor background worker task was added to periodically watch
the cgroup cpu/memory limit setting changes. Metrics will be set
accordingly and displayed on Grafana. New alerting rules were also added
for the cpu/memory usage ratio relative to the limits.
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [x] Manual test (add detailed scripts or steps below)
        Grafana display and alerting trigger
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
Monitoring Cgroup CPU and memory limit change.
```

### CODE REVIEW

**Comment by nolouch** [comment]
Thank you for your contribution. Could you please fix the CI test?

**Review by overvenus — APPROVED** [review]
LGTM, thanks!

File: metrics/grafana/tikv_summary.json [hunk 1]
```diff
@@ -500,6 +508,14 @@
               "legendFormat": "{{instance}}",
               "refId": "A",
               "step": 10
+            },
+            {
+              "expr": "tikv_server_memory_quota_bytes{k8s_cluster=\"$k8s_cluster\", tidb_cluster=\"$tidb_cluster\", instance=~\"$instance\", job=~\".*tikv\"}",
```
Comment (Connor1996) [inline metrics/grafana/tikv_summary.json:513]: should it be added to `tikv_details` as well?

**Review by Connor1996 — COMMENTED** [review]
rest LGTM

File: metrics/grafana/tikv_summary.json (hunk 1 above)
Comment (pingandb) [reply metrics/grafana/tikv_summary.json:513]: > should it be added to `tikv_details` as well?

Done.

**Review by Connor1996 — APPROVED** [review]
LGTM

**Comment by pingandb** [comment]
> Please fix the conflict

Tried rebasing with master and found the work tree clean. Any advice?

**Comment by Connor1996** [comment]
> > Please fix the conflict
> 
> Tried rebasing with master and found the working tree clean. Any advice?

What do you mean by "working tree clean"

**Comment by pingandb** [comment]
> > > Please fix the conflict
> > 
> > 
> > Tried rebasing with master and found the working tree clean. Any advice?
> 
> What do you mean by "working tree clean"

image

**Comment by Connor1996** [comment]
Regenerate the dashboard, then push

**Comment by pingandb** [comment]
> Regenerate the dashboard, then push

I did make the change by modifying metrics/grafana/tikv_details.dashboard.py and then running/scripts/gen-tikv-details-dashboard to generate tikv_details.json and tikv_details.json.sha256. The branch was already up to date when the files generated. So I don't think regenerating the file can solve the problem. And on the other hand, the script won't regenerate the files when no changes detected.

**Comment by pingandb** [comment]
> PR needs rebase.

It was already rebased.

**Comment by overvenus** [comment]
/merge

Thank you! @pingandb

**Pull request merged**